Add feature(nll)

spastorino opened the pull request:
This is for https://github.com/rust-lang/rust/issues/46716

nikomatsakis committed:
add some run-pass tests for NLL showing that things work as expected

nikomatsakis commented on src/librustc/session/mod.rs:
```diff
@@ -437,6 +437,12 @@ impl Session {
+        self.features.borrow().nll || self.opts.debugging_opts.nll
+    }
+    pub fn nll_dump_cause(&self) -> bool {
+        self.features.borrow().nll || self.opts.debugging_opts.nll_dump_cause
```
I am debating whether we should have `#![feature(nll)]` enable dump-cause or not. The main reason not to is that it is **so sloooooow** right now. I feel like we might want to enable it once we make it faster (and then we wouldn't have this switch at all.)

nikomatsakis reviewed (commented):
Looks basically right, modulo the question about `-Znll-dump-cause`, and the fact that it needs to handle `-Zborrowck=mir`

spastorino committed:
Add nll_dump_cause helper to Session

spastorino committed:
Add nll feature and make nll imply nll_dump_cause

spastorino committed:
feature nll implies two-phase-borrows

spastorino committed:
feature nll implies borrowck=mir

spastorino commented:
This should be ready!

nikomatsakis committed:
document and tweak the nll, use_mir, etc helpers

In particular, -Znll might as well imply -Zborrowck=mir by default,
just like `#![feature(nll)]` does.

Also, if NLL is in use, no reason to emit end regions. The NLL pass
just strips them out anyway.

nikomatsakis committed:
when using feature(nll), don't warn about AST-based region errors

Also, keep reporting AST-based region errors that are not occuring in
a fn body.
